mono/metadata/native-library.c

```diff
@@ -408,126 +392,393 @@ mono_lookup_pinvoke_call_internal (MonoMethod *method, MonoError *error)
+
+	mono_error_cleanup (error);
+
+	return result;
+}
+
+static MonoDl *
+netcore_lookup_native_library (MonoAssemblyLoadContext *alc, MonoImage *image, const char *scope, guint32 flags)
+{
+	MonoDl *module = NULL;
+	MonoAssembly *assembly = mono_image_get_assembly (image);
+	char *error_msg = NULL;
+
+	MONO_REQ_GC_UNSAFE_MODE;
+
+	mono_trace (G_LOG_LEVEL_INFO, MONO_TRACE_DLLIMPORT, "DllImport attempting to load: '%s'.", scope);
+
+	// We allow a special name to dlopen from the running process namespace, which is not present in CoreCLR
+	if (strcmp (scope, "__Internal") == 0) {
+		if (!internal_module)
+			internal_module = mono_dl_open (NULL, MONO_DL_LAZY, &error_msg);
+		module = internal_module;
+
+		if (!module) {
+			mono_trace (G_LOG_LEVEL_INFO, MONO_TRACE_DLLIMPORT, "DllImport error loading library '__Internal': '%s'.", error_msg);
+			g_free (error_msg);
 		}
-		mono_image_unlock (image);
+
+		return module;
 	}
 
-	mono_trace (G_LOG_LEVEL_INFO, MONO_TRACE_DLLIMPORT,
-				"DllImport searching in: '%s' ('%s').", new_scope, found_name);
+	/*
+	 * Try these until one of them succeeds:
+	 *
+	 * 1. Check the cache in the active ALC.
+	 *
+	 * 2. Call the DllImportResolver on the active assembly.
+	 *
+	 * 3. Call LoadUnmanagedDll on the active ALC.
+	 *
+	 * 4. Check the global cache.
+	 *
+	 * 5. Run the unmanaged probing logic.
+	 *
+	 * 6. Raise the ResolvingUnmanagedDll event on the active ALC.
+	 *
+	 * 7. Return NULL.
+	 */
 
-	addr = pinvoke_probe_for_symbol (module, piinfo, new_import, &error_msg);
+	mono_alc_pinvoke_lock (alc);
+	module = (MonoDl *)g_hash_table_lookup (alc->pinvoke_scopes, scope);
+	native_library_lock ();
+	if (netcore_check_blacklist (module)) {
+		g_hash_table_remove (alc->pinvoke_scopes, scope);
+		module = NULL;
+	}
+	native_library_unlock ();
+	mono_alc_pinvoke_unlock (alc);
+	if (module)
+		goto leave;
 
-	if (!addr) {
-		status_out->err_code = LOOKUP_PINVOKE_ERR_NO_SYM;
-		status_out->err_arg = g_strdup (new_import);
+	module = (MonoDl *)netcore_resolve_with_dll_import_resolver_nofail (alc, assembly, scope, flags);
+	if (module)
+		goto add_to_alc_cache;
+
+	module = (MonoDl *)netcore_resolve_with_load_nofail (alc, scope);
+	if (module)
+		goto add_to_alc_cache;
+
+	MONO_ENTER_GC_SAFE;
+	mono_global_loader_data_lock ();
+	module = (MonoDl *)g_hash_table_lookup (global_module_map, scope);
+	mono_global_loader_data_unlock ();
+	MONO_EXIT_GC_SAFE;
+	if (module)
+		goto add_to_alc_cache;
+
+	module = netcore_probe_for_module (image, scope, flags);
+	if (module)
+		goto add_to_global_cache;
+
+	// As this is last chance, I've opted not to put it in a cache, but that is not necessarily the correct decision.
+	module = netcore_resolve_with_resolving_event_nofail (alc, assembly, scope);
+	goto leave;
+
+add_to_global_cache:
+	MONO_ENTER_GC_SAFE;
+	mono_global_loader_data_lock ();
+	mono_loader_register_module (scope, module);
+	mono_global_loader_data_unlock ();
+	MONO_EXIT_GC_SAFE;
```

Comment:
1. `mono_loader_register_module` already has code to take the global loader data lock.

2. I don't think you want to be in GC Safe for the duration of `mono_loader_register_module` - you want two transitions - one around locking and one around unlocking.